rust/azure_iot_operations_services/src/leased_lock/lease.rs

```diff
@@ -0,0 +1,416 @@
+    ///
+    /// [`struct@Error`] of kind [`UnexpectedPayload`](ErrorKind::UnexpectedPayload) if the State Store returns a response that isn't valid for the request
+    ///
+    /// [`struct@Error`] of kind [`AIOProtocolError`](ErrorKind::AIOProtocolError) if there are any underlying errors from the command invoker
+    /// # Panics
+    /// Possible panic if, for some error, the fencing token (a.k.a. `version`) of the acquired lease is None.
+    pub async fn acquire(
+        &mut self,
+        lease_expiration: Duration,
+        request_timeout: Duration,
+        renewal_period: Option<Duration>,
+    ) -> Result<HybridLogicalClock, Error> {
+        // Logic:
+        // a. Start observing lease within this function.
+        // b. Try acquiring the lease and return if acquired or got an error other than `KeyAlreadyLeased`.
+        // c. If got `KeyAlreadyLeased`, wait until `Del` notification for the lease. If notification is None, re-observe (start from a. again).
+        // d. Loop back starting from b. above.
+        // e. Unobserve lease before exiting.
+
+        self.stop_lease_auto_renewal().await;
+
+        let mut observe_response = self.observe(request_timeout).await?;
+        let mut acquire_result;
+
+        loop {
+            acquire_result = self.try_acquire(lease_expiration, request_timeout).await;
+
+            match acquire_result {
+                Ok(_) => {
+                    break; /* lease acquired */
+                }
+                Err(ref acquire_error) => match acquire_error.kind() {
+                    ErrorKind::KeyAlreadyLeased => { /* Must wait for lease to be released. */ }
+                    _ => {
+                        break;
+                    }
+                },
+            };
+
+            // Lease being held by another client. Wait for delete notification.
+            loop {
+                let Some((notification, _)) = observe_response.response.recv_notification().await
+                else {
+                    // If the state_store client gets disconnected (or shutdown), all the observation channels receive a None.
+                    // In such case, as per design, we must re-observe the lease.
+                    observe_response = self.observe(request_timeout).await?;
+                    break;
+                };
+
+                if notification.operation == state_store::Operation::Del {
+                    break;
+                };
+            }
+        }
+
+        match self.unobserve(request_timeout).await {
+            Ok(_) => {
+                if acquire_result.is_ok() {
+                    self.auto_renewal_task = Self::start_lease_auto_renewal(
+                        self.state_store.clone(),
+                        self.key_name.clone(),
+                        self.lease_holder_name.clone(),
+                        lease_expiration,
+                        request_timeout,
+                        renewal_period,
```

Comment:
I think instead of passing in renewal_period as an option here and potentially returning a dummy task, you could instead only call start_lease_auto_renewal if renewal_period is Some